### PR TITLE
feat: Pull podman images and create networks in preparation

### PR DESCRIPTION
**Pull request by Tiboris**
Pull podman image in provisioning preparation
to not block the provisioning or pull one image
more than once which may cause unexpected errors.

Create podman networks in provisioning preparation
to not block the provisioning or ty to create already
created networks which may cause unexpected errors.

Signed-off-by: Tibor Dudlák <tdudlak@redhat.com>

### CODE REVIEW

File: src/mrack/providers/podman.py [hunk 1]
```diff
@@ -58,8 +58,33 @@ async def init(self, container_images, default_network, ssh_key, container_optio
...
+        image_list.add("quay.io/neoave/snappeas:fedora-33")
+        for req in reqs:
+            image = req["image"]
+            if not await self.podman.image_exists(image):
+                logger.debug(f"{self.dsp_name}: Pull of image '{image}' required")
+                image_list.add(image)
+            else:
+                logger.debug(f"{self.dsp_name}: Image '{image}' present")
+
+        if not image_list:
```
Comment (pvoborni) [inline src/mrack/providers/podman.py:72]: The success message that we have all the images could be moved after the `success = all(pull_results)`.  So that there would be some pointer that this part of operation ended.

File: src/mrack/providers/podman.py [hunk 2]
```diff
@@ -58,8 +58,33 @@ async def init(self, container_images, default_network, ssh_key, container_optio
         logger.info(f"{self.dsp_name}: Init duration {login_duration}")
 
     async def prepare_provisioning(self, reqs):
-        """Prepare provisioning."""
-        pass
+        """Pull missing images to prepare provisioning for podman."""
+        image_list = set()
+        image_list.add("quay.io/neoave/snappeas:fedora-33")
```
Comment (pvoborni) [inline src/mrack/providers/podman.py:63]: Is having ` image_list.add("quay.io/neoave/snappeas:fedora-33")` here intentional? Looks as leftover from testing.

**Review by pvoborni — COMMENTED** [review]
Thanks for the PR. Some comments added. 

A minor thing for next time: it would also good if the unrelated changes of logging messages would be in a separate commit.

File: src/mrack/providers/podman.py (hunk 2 above)
Comment (Tiboris) [reply src/mrack/providers/podman.py:63]: Correct

File: src/mrack/providers/podman.py (hunk 1 above)
Comment (Tiboris) [reply src/mrack/providers/podman.py:72]: Can do

File: src/mrack/providers/podman.py [hunk 3]
```diff
@@ -58,17 +58,42 @@ async def init(self, container_images, default_network, ssh_key, container_optio
...
 
         for req in reqs:
+            # First create list of podman network(s) to be created
+            # podman network(s) should be always created no matter what
+            network = req.get(
+                "network", f"{self.default_network}-{req['domain'].replace('.', '-')}"
+            )
+            # store the used network per requirement to be later accessed
+            req["network"] = network
+            if not await self.podman.network_exists(network):
```
Comment (pvoborni) [inline src/mrack/providers/podman.py:73]: `self.podman.network_exists(network)` check is run for all reqs no matter what in sequence. So e.g. for 5 machines with the same network, it is run 4 times unnecessarily. 

Better logic would be:
- generate the list of network names in `network_list` (and as it is a set, it will have only unique)
- run `self.podman.network_exists(network):` for each network in network_list
- create the ones which doesn't exist

Comment (Tiboris) [reply src/mrack/providers/podman.py:73]: Correct, will update.

Comment (Tiboris) [reply src/mrack/providers/podman.py]: Done

File: src/mrack/providers/podman.py [hunk 4]
```diff
@@ -58,8 +58,58 @@ async def init(self, container_images, default_network, ssh_key, container_optio
...
+            network = req.get(
+                "network", f"{self.default_network}-{req['domain'].replace('.', '-')}"
+            )
+            # store the used network per requirement to be later accessed
+            req["network"] = network
+            network_list.add(network)
+
+            # Prepare image_list to be pulled later
+            image = req["image"]
+            if not await self.podman.image_exists(image):
```
Comment (pvoborni) [inline src/mrack/providers/podman.py:77]: This 'if' should be on line 102 as I wrote earlier.  Where it woudl check it only once per image. 

So that line 100 would be like:
```python
        awaitables = []
        for image in image_list:
            if not await self.podman.image_exists(image):
                 awaitables.append(self.podman.pull(image))
```

Comment (Tiboris) [reply src/mrack/providers/podman.py]: Updated

**Comment by bhavikbhavsar** [comment]
Following scenarios where tested:

1.  Defined multiple hosts in the metadata for different OS, pre-setup pull only one image.
https://pastebin.com/UhAqh9rZ - success
Only missing image is pulled.
Hosts within the same domain are able to communicate.
Qs:
 - what if user wants to have the latest image pulled whenever mrack provisions?

2. When url for image is incorrect in provisioning-config
https://pastebin.com/j0hZ21fv - success
Proper error is displayed

3. When multiple domains defined in metadata.
https://pastebin.com/9aG8DrKz - Fails
Fails after creating the first two host from first domain as hostname are treated as same.
Qs:
- What should be default behaviour here, is it necessary for user to define separate hostname?

4. Define different hostname for each domain.
https://pastebin.com/YkdZATHv 
Current behaviours different network ips are assigned as per the domains. But all provisioned containers are able to communicate with one another.
Qs:
- What is the expected behaviour, when provisioned in two different domains? Should it be able to communicate between different domains or it should not?

**Comment by pvoborni** [comment]
> Following scenarios were tested:
> 
>     1. Defined multiple hosts in the metadata for different OS, pre-setup pull only one image.
>        https://pastebin.com/UhAqh9rZ - success
>        Only missing image is pulled.
>        Hosts within the same domain are able to communicate.
>        Qs:
> 
> 
>     * what if user wants to have the latest image pulled whenever mrack provisions?

Don't know if we need it. But can be an RFE if needed.

> 
> 
>     1. When url for image is incorrect in provisioning-config
>        https://pastebin.com/j0hZ21fv - success
>        Proper error is displayed
> 
>     2. When multiple domains defined in metadata.
>        https://pastebin.com/9aG8DrKz - Fails
>        Fails after creating the first two host from first domain as hostname are treated as same.
>        Qs:
> 
> 
>     * What should be default behaviour here, is it necessary for user to define separate hostname?

It is expected in Mrack that each name of a host will be unique, regardless of domain. In such case it is better to use FQDNs. I don't know what is the error in such case, but it might be good RFE to fail early in such situation. In any case it is IMO not related to this PR. 

> 
> 
>     1. Define different hostname for each domain.
>        https://pastebin.com/YkdZATHv
>        Current behaviours different network ips are assigned as per the domains. But all provisioned containers are able to communicate with one another.
>        Qs:
> 
> 
>     * What is the expected behaviour, when provisioned in two different domains? Should it be able to communicate between different domains or it should not?

They should be able to communicate. E.g. for testing cross realm trusts (not supported now in IPA, but an example).

> 
> 
> Also a general observation between each mrack up had to delete mrackdb.json. It becomes annoying. Can a hashmap be created to record each run in db? So each time db file need not be deleted.

I'd rephrase that if in the **same** working directory, host definition in **metadata changes**, especially in names of hosts, then it is required to delete the DB file. It is not required if metadata doesn't change.  I guess we can define, what would be a good behavior in such case. In either supporting it better or at least writing better errors. That said, I'd recommend to have a separate working dir for different test cases.

**Comment by Tiboris** [comment]
> > Following scenarios were tested:
> > ```
> > 1. Defined multiple hosts in the metadata for different OS, pre-setup pull only one image.
> >    https://pastebin.com/UhAqh9rZ - success
> >    Only missing image is pulled.
> >    Hosts within the same domain are able to communicate.
> >    Qs:
> > 
> > 
> > * what if user wants to have the latest image pulled whenever mrack provisions?
> > ```
> 
> Don't know if we need it. But can be an RFE if needed.

I like this idea, probably some option in provisioning-config.yaml to re pull if image exists.

**Review by pvoborni — APPROVED** [review]
Code looks good to me. Approving - referring to Bhavik's testing.
